xlators/cluster/afr/src/afr-common.c

```diff
@@ -5129,11 +5137,14 @@ afr_lk_cbk(call_frame_t *frame, void *cookie, xlator_t *this, int32_t op_ret,
     child_index++;
 
     if (child_index < priv->child_count) {
-        STACK_WIND_COOKIE(frame, afr_lk_cbk, (void *)(long)child_index,
-                          priv->children[child_index],
-                          priv->children[child_index]->fops->lk, local->fd,
-                          local->cont.lk.cmd, &local->cont.lk.user_flock,
-                          local->xdata_req);
+        local->child_index = child_index;
+        ret = synctask_new(this->ctx->env, afr_lk_txn, afr_lk_txn_cbk, frame,
```

Comment:
This will create nested synctasks for each subvolume. I'm not sure if it's a bit overkill...

Reply:
> Sorry for the slow reviews... I've just commented on two things, but otherwise the change seems right.
> 
> Given that there doesn't seem to be any objections, I'll probably merge it when regression run finishes, and we'll check if the nightly performance check shows any regression as a consequence of the synctasks. If not. we are fine. Otherwise we'll need to address the performance issue in another patch.

I have some time this Friday. Will do this review.

Reply:
@xhernandez One doubt, if the fd is open and it loses locks because of loss of connection, doesn't that mean the fd is bad? Why should we re-open if we can't heal locks? Application should get errors in this situation right? So the existing behavior seems correct to me. If we can heal locks, then re-opening makes sense. But at the moment, it doesn't heal locks.

Reply:
If the fd is not bad, we do not reopen. Only when the file is not opened or the fd is bad, we reopen it. So do we  need to heal lock?
Please review it again. @pranithk @xhernandez 
Thank you.

Reply:
I understand, but the existing behavior is correct unless we plan to implement lock healing. This fix I think will lead to applications thinking they still have lock while it lost them.
Consider the scenario:
1. client-1 took full lock on a.txt
2. client-2 is blocked on full lock on a.txt
3. cleint-1 experienced disconnects to all bricks
4. client-2 may have gotten the lock and modified the file
5. Now when client-1 is reconnected, operations on the fd with a.txt should fail because the locks on that fd are not valid anymore. Right?

Reply:
> @xhernandez One doubt, if the fd is open and it loses locks because of loss of connection, doesn't that mean the fd is bad? Why should we re-open if we can't heal locks? Application should get errors in this situation right? So the existing behavior seems correct to me. If we can heal locks, then re-opening makes sense. But at the moment, it doesn't heal locks.

I think you are right. This approach is not enough to prevent some dangerous cases. However, I don't think the current approach is correct either:

1. After first disconnect and reconnect, without any other issue in the volume, self-heal may be continuously active repairing files that cannot be updated in the normal I/O path because of the missing fd, wasting resources and reducing performance.
2. When too many bricks have disconnected, applications will start receiving massive EBADF errors, which seems a misleading error to me, specially when everything should be fine because the volume has never gone below the required quorum.

Shouldn't we implement posix lock recovery ? something very similar is already implemented for graph switch. Couldn't we improve it to also work on reconnections taking the data from other bricks ?

Reply:
Agree with 1.
Not sure I understand 2. Once the fd is marked bad, it is an unrecoverable error with some sort of lock healing. I don't know how a lock healing can be attempted if disconnect leads to loss of locks, so you can never trust this fd anymore. The only way to recover is if we have some sort of grace-period in which we allow the locks to be recovered. But it may need persistence of locks too if the disconnects are happening because of brick processes dying and coming back up. So neither a small nor an easy fix?

Reply:
> Agree with 1. Not sure I understand 2. Once the fd is marked bad, it is an unrecoverable error with some sort of lock healing. I don't know how a lock healing can be attempted if disconnect leads to loss of locks, so you can never trust this fd anymore. The only way to recover is if we have some sort of grace-period in which we allow the locks to be recovered. But it may need persistence of locks too if the disconnects are happening because of brick processes dying and coming back up. So neither a small nor an easy fix?

As long as there are not too many simultaneous disconnections we could regenerate the posix locks on one brick from the data on another healthy brick. I haven't thought about all the details, though, so I could be missing something important, but it seems to me that it should be doable.

Reply:
Okay, will be happy to check if the solution meets the requirements.

Reply:
> > Agree with 1. Not sure I understand 2. Once the fd is marked bad, it is an unrecoverable error with some sort of lock healing. I don't know how a lock healing can be attempted if disconnect leads to loss of locks, so you can never trust this fd anymore. The only way to recover is if we have some sort of grace-period in which we allow the locks to be recovered. But it may need persistence of locks too if the disconnects are happening because of brick processes dying and coming back up. So neither a small nor an easy fix?
> 
> As long as there are not too many simultaneous disconnections we could regenerate the posix locks on one brick from the data on another healthy brick. I haven't thought about all the details, though, so I could be missing something important, but it seems to me that it should be doable.

@xhernandez  In a world, what is the idea of our fix？ 
Please tell me it again. Thank you!

Reply:
> > Agree with 1. Not sure I understand 2. Once the fd is marked bad, it is an unrecoverable error with some sort of lock healing. I don't know how a lock healing can be attempted if disconnect leads to loss of locks, so you can never trust this fd anymore. The only way to recover is if we have some sort of grace-period in which we allow the locks to be recovered. But it may need persistence of locks too if the disconnects are happening because of brick processes dying and coming back up. So neither a small nor an easy fix?
> 
> As long as there are not too many simultaneous disconnections we could regenerate the posix locks on one brick from the data on another healthy brick. I haven't thought about all the details, though, so I could be missing something important, but it seems to me that it should be doable.

@xhernandez What should we do next to fix the problem.
Please tell me it.
Thank you.

Reply:
> @xhernandez What should we do next to fix the problem. Please tell me it. Thank you.

The change requires significantly more work if you are also going to fix the posix locks. My idea is to reuse (or adapt if necessary) a feature already used during graph switch to move the posix locks from one brick to another. It's implemented in fuse xlator inside `fuse_migrate_locks()`. It uses a special xattr to read lock info from one brick and move it to the new brick. I think we should do something similar inside AFR just after reopening the fd, reading the posix lock information from one healthy brick and sending it to the newly opened fd in another brick. However I haven't checked the code in depth and I'm not sure if it can be directly reused for this purpose or it requires some changes.

Reply:
> > @xhernandez What should we do next to fix the problem. Please tell me it. Thank you.
> 
> The change requires significantly more work if you are also going to fix the posix locks. My idea is to reuse (or adapt if necessary) a feature already used during graph switch to move the posix locks from one brick to another. It's implemented in fuse xlator inside `fuse_migrate_locks()`. It uses a special xattr to read lock info from one brick and move it to the new brick. I think we should do something similar inside AFR just after reopening the fd, reading the posix lock information from one healthy brick and sending it to the newly opened fd in another brick. However I haven't checked the code in depth and I'm not sure if it can be directly reused for this purpose or it requires some changes.

@xhernandez  One doubt, If there is a volume with two copies, there may not be a healthy brick.

Reply:
> > The change requires significantly more work if you are also going to fix the posix locks. My idea is to reuse (or adapt if necessary) a feature already used during graph switch to move the posix locks from one brick to another. It's implemented in fuse xlator inside `fuse_migrate_locks()`. It uses a special xattr to read lock info from one brick and move it to the new brick. I think we should do something similar inside AFR just after reopening the fd, reading the posix lock information from one healthy brick and sending it to the newly opened fd in another brick. However I haven't checked the code in depth and I'm not sure if it can be directly reused for this purpose or it requires some changes.
> 
> @xhernandez One doubt, If there is a volume with two copies, there may not be a healthy brick.

If you mean replica 2, unless there's something that clearly indicates that the "alive" brick is bad, we consider it as healthy. Otherwise we can simply skip reopening the fd as we do now. The limitations of replica 2 have been explained from long ago. Gluster will do its best, but it's easy to have inconsistencies when there are disconnections.

In a replica 3 all this is easier and safer to track.